Rescale ODEs to match Euler solver

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->

Fixes #15217 

Needs a fix for #15433 for one of the tests

#### Brief description of what is fixed or changed

When matching an ODE against the Euler form this patch multiplies the coefficients by an appropriate power of `x`. This means that `dsolve` can recognise more ODEs as being Euler ODEs and can solve some ODEs that it couldn't before.

This example gives `NotImplementedError` on current master:
```julia
In [7]: eqn = -exp(x) + (x*Derivative(f(x), (x, 2)) + Derivative(f(x), x))/x

In [8]: eqn
Out[8]: 
           2                 
          d          d       
       x⋅───(f(x)) + ──(f(x))
           2         dx      
   x     dx                  
- ℯ  + ──────────────────────
                 x           

In [9]: dsolve(eqn, f(x))
Out[9]: 
                                            ⌠               
                                 x          ⎮    x          
f(x) = C₁ + C₂⋅log(x) + (x - 1)⋅ℯ ⋅log(x) - ⎮ x⋅ℯ ⋅log(x) dx
                                            ⌡    
```

#### Other comments

I've had to adjust the tests in `classify_ode` because the Euler solver now matches many trivial ODEs.

I found #15433 while working on this. One of the tests uses an awkward workaround for that. Probably #15433 should be fixed first.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* solvers
    * dsolve now rescales ODEs so that the Euler solver can be used in more cases
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
Rescale ODEs to match Euler solver

oscarbenjamin committed:
Update classify_ode doctest after nth_euler fixes

sidhantnagpal commented on sympy/solvers/tests/test_ode.py:
```diff
@@ -2420,6 +2428,16 @@ def test_nth_order_linear_euler_eq_nonhomogeneous_variation_of_parameters():
     assert checkodesol(eq, sol, order=2, solve_for_func=False)[0]
 
+    eq = -exp(x) + (x*Derivative(f(x), (x, 2)) + Derivative(f(x), x))/x
+    sol = Eq(f(x), C1 + C2*log(x) + (x - 1)*exp(x)*log(x) - Integral(x*exp(x)*log(x), x))
```
The following will work without calling `.replace()`:
```
    eq = -exp(x) + (x*Derivative(f(x), (x, 2)) + Derivative(f(x), x))/x
    from sympy.integrals.risch import NonElementaryIntegral
    sol = Eq(f(x), C1 + C2*log(x) + (x - 1)*exp(x)*log(x) - NonElementaryIntegral(x*exp(x)*log(x), x))
    assert our_hint in classify_ode(eq)
    assert dsolve(eq, f(x), hint=our_hint) == sol
    assert checkodesol(eq, sol, order=2, solve_for_func=False)[0]
```

oscarbenjamin committed:
Use NonElementaryIntegral in test solution

oscarbenjamin commented:
I've just changed the test so that the expected solution is in terms of `NonElementaryIntegral`. I'm not totally satisfied with that test since it is actually possible to get that solution in a better form using `Ei(x)` but `integrate` can't do it yet: #15431 

Otherwise I'm happy with this PR now.

asmeurer commented:
Why not make the ODE monic, as I suggested in the issue. That would also handle other things, like the whole ODE being multiplied by some function.

oscarbenjamin committed:
Match Euler ODEs multiplied through by a function

oscarbenjamin commented:
I've changed it so that it makes the equation *almost* monic. It multiples through by `x**n/a` where a is the leading coefficient. This puts into something like the monic form of an Euler ODE.

I've also added a test for the case that the ODE is multiplied through by a function of `x`.

asmeurer commented:
I see. I thought it was only detecting constant `a`.

oscarbenjamin commented:
It was - I just pushed another commit before my previous comment.
